### PR TITLE
Add AllocatorState

### PR DESCRIPTION
**Pull request by nordlow**
For sample usage see

https://github.com/nordlow/phobos-next/blob/master/src/nxt/dynamic_array.d#L828

Also see discussion at https://forum.dlang.org/post/uiavxptceyxsjulkxlec@forum.dlang.org.

Depends on https://github.com/dlang/phobos/pull/8598.

### CODE REVIEW

File: std/experimental/allocator/common.d [hunk 1]
```diff
@@ -11,6 +11,28 @@ Source: $(PHOBOSSRC std/experimental/allocator/common.d)
 module std.experimental.allocator.common;
 import std.algorithm.comparison, std.traits;
 
+/** Is `true` iff `A` is an allocator.
+ *
+ * Only members and `alignment` and `allocate` are required.
+ */
+enum isAllocator(A) = (is(typeof(A.allocate(size_t.init)) == void[])
+					   && is(typeof(A.alignment) : size_t)); // currently either uint or size_t.
```
Comment (nordlow) [inline std/experimental/allocator/common.d:19]: Shall we require `A.alignment` to be known at compile-time?

File: std/experimental/allocator/common.d [hunk 2]
```diff
@@ -11,6 +11,28 @@ Source: $(PHOBOSSRC std/experimental/allocator/common.d)
 module std.experimental.allocator.common;
 import std.algorithm.comparison, std.traits;
 
+/** Is `true` iff `A` is an allocator.
+ *
+ * Only members and `alignment` and `allocate` are required.
+ */
+enum isAllocator(A) = (is(typeof(A.allocate(size_t.init)) == void[])
+					   && is(typeof(A.alignment) : size_t));
```
Comment (nordlow) [inline std/experimental/allocator/common.d:19]: Shall we require `A.alignment` to be known at compile-time?

**Comment by nordlow** [comment]
> I don't think that's needed, there is no point adding more code there only just to satisfy the need of a 3rd party library, they should maintain that code

I believe `isAllocator` is of higher importance than `AllocatorState`.

**Comment by atilaneves** [comment]
Two questions:

1. What is this for?
2. Assuming there's a use for this, why wouldn't be an enum called `Size`?

**Comment by nordlow** [comment]
> 1. What is this for?

Some allocators need to hold state others don't this inlines that static if construct as shown at

https://github.com/nordlow/phobos-next/blob/master/src/nxt/dynamic_array.d#L828

**Comment by atilaneves** [comment]
Ah, I guess it's the equivalent of abstracting [this](https://github.com/atilaneves/automem/blob/87fd0106a1d95edd0f21625d97b30e44d77483f1/source/automem/vector.d#L473)?

**Comment by nordlow** [comment]
> Ah, I guess it's the equivalent of abstracting [this](https://github.com/atilaneves/automem/blob/87fd0106a1d95edd0f21625d97b30e44d77483f1/source/automem/vector.d#L473)?

Yep.

**Comment by nordlow** [comment]
> Ah, I guess it's the equivalent of abstracting [this](https://github.com/atilaneves/automem/blob/87fd0106a1d95edd0f21625d97b30e44d77483f1/source/automem/vector.d#L473)?

Btw, you don't need the do keyword in places such as

https://github.com/atilaneves/automem/blob/87fd0106a1d95edd0f21625d97b30e44d77483f1/source/automem/vector.d#L310

, @atilaneves.

**Comment by atilaneves** [comment]
Users could want the allocator member as a pointer. I don't think this is general enough to merit inclusion into Phobos, ideally most of the code using allocators directly would *be* Phobos code.

**Comment by nordlow** [comment]
> I don't think that's needed, there is no point adding more code there only just to satisfy the need of a 3rd party library, they should maintain that code

This pattern applies to all containers taking an arbitrary allocator as parameter. That's more generic than a a single 3:rd party library.

**Comment by atilaneves** [comment]
> > I don't think that's needed, there is no point adding more code there only just to satisfy the need of a 3rd party library, they should maintain that code
> 
> `AllocatorState` can be used in every container utilizing `std.experimental.allocator`. That's way more general than a single 3:rd party library.

I don't think it is. For instance, how would I use that to hold the allocator as a pointer?

**Comment by nordlow** [comment]
> I don't think it is. For instance, how would I use that to hold the allocator as a pointer?

Good point. But how often do we need to model an allocator state as pointer?

Ping, @andralex.

**Comment by atilaneves** [comment]
I don't think this is general enough to warrant inclusion in Phobos.

**Comment by nordlow** [comment]
I've extended `AllocatorState` to support `Allocator` being a pointer to an Allocator

https://github.com/nordlow/phobos-next/blob/06928467993991e057a2b04bd7a6dbbad5dcc6d7/src/nxt/allocator_traits.d#L36

. Of course we could want to support arbitrary number pointer of indirections...

**Comment by RazvanN7** [comment]
@nordlow It seems that @atilaneves does not approve of this addition on the basis that having an allocator that requires a state is not something frequently seen. I stand by this point as this is something library creators can handle on their own if the need arises.

**Comment by nordlow** [comment]
> I don't think this is general enough to warrant inclusion in Phobos.

Can you explain how you could make it more general?

**Comment by atilaneves** [comment]
I think it's not "general enough" in the sense that this will see little to no usage, and that said usage should be limited to smart pointers / container types anyway.

**Comment by nordlow** [comment]
> having an allocator that requires a state is not something frequently seen. 

Do we have any statistics backing this claim? This MR is part of a vision I have for making allocators more integrated into std.experimental.allocator Phobos and third-party containers.

Ping, @andralex.

**Review by thewilsonator — APPROVED** [review]

**Review by LightBender — APPROVED** [review]

**Comment by nordlow** [comment]
I have a refined version that also treats to pointers to allocators as state as

```d
mixin template AllocatorState(Allocator) /+ TODO: add string fieldName parameter +/
if (isAllocator!Allocator ||
	isAllocator!(typeof(*Allocator.init))) {
	private import std.experimental.allocator.common : stateSize;
	static if (stateSize!Allocator == 0)
		alias allocator = Allocator.instance;
	else
		Allocator allocator;
}

version(nxt_test) version(unittest) {
	pure nothrow @safe @nogc version(nxt_test) unittest
	{
		mixin AllocatorState!NullAllocator n;
		mixin AllocatorState!GCAllocator g;
		mixin AllocatorState!Mallocator m;
		mixin AllocatorState!MmapAllocator p;
		mixin AllocatorState!(MmapAllocator*) pi;
	}
}

version(nxt_test) version(unittest) {
	import std.experimental.allocator.building_blocks.null_allocator : NullAllocator;
	import std.experimental.allocator.mallocator : Mallocator;
	import std.experimental.allocator.gc_allocator : GCAllocator;
	import std.experimental.allocator.mmap_allocator : MmapAllocator;
}
```

. Shall I add this aswell?

**Comment by LightBender** [comment]
Sure. You are more than welcome to bring new improvements to this process.

**Comment by nordlow** [comment]
> Sure. You are more than welcome to bring new improvements to this process.

Let's merge this as is. If I see a need for having the allocator be passed as a pointer I'll add that then. Ok?

**Comment by LightBender** [comment]
Whatever works best for you. Can you rebase or push a commit to trigger the testers? Otherwise it might be a few hours.

**Comment by nordlow** [comment]
> Whatever works best for you. Can you rebase or push a commit to trigger the testers? Otherwise it might be a few hours.

Done.

**Comment by nordlow** [comment]
MacOS build error at https://github.com/dlang/phobos/actions/runs/11539950157/job/32120212715?pr=8596#step:12:132 seems unrelated.

**Comment by thewilsonator** [comment]
Wait for the two unfinished checks. They are kinda expected to fail randomly at the moment.

**Pull request merged**